pkg/suggestion/suggest.go

```diff
@@ -0,0 +1,137 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package suggestion
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+// SubcommandsRequiredWithSuggestions will ensure we have a subcommand provided by the user and augments it with
+// suggestion for commands, alias and help on root command.
+func SubcommandsRequiredWithSuggestions(cmd *cobra.Command, args []string) error {
+	requireMsg := "unknown command \"%s\" for \"%s\""
+	typedName := ""
+	// This will be triggered if cobra didn't find any subcommands.
+	// Find some suggestions.
+	var suggestions []string
+
+	if len(args) != 0 && !cmd.DisableSuggestions {
+		typedName += args[0]
+		if cmd.SuggestionsMinimumDistance <= 0 {
+			cmd.SuggestionsMinimumDistance = 2
+		}
+		// subcommand suggestions
+		suggestions = cmd.SuggestionsFor(args[0])
+
+		// subcommand alias suggestions (with distance, not exact)
+		for _, c := range cmd.Commands() {
+			if c.IsAvailableCommand() {
+				candidate := suggestsByPrefixOrLd(typedName, c.Use, cmd.SuggestionsMinimumDistance)
+				if candidate == "" {
+					continue
+				}
+				_, found := Find(suggestions, candidate)
+				if !found {
+					suggestions = append(suggestions, candidate)
+				}
+			}
+		}
+
+		// help for root command
+		if !cmd.HasParent() {
+			candidate := suggestsByPrefixOrLd(typedName, "help", cmd.SuggestionsMinimumDistance)
+			if candidate != "" {
+				suggestions = append(suggestions, candidate)
+			}
+		}
+	}
+
+	var suggestionsMsg string
+	if len(suggestions) > 0 {
+		suggestionsMsg += "\nDid you mean this?\n"
+		for _, s := range suggestions {
+			suggestionsMsg += fmt.Sprintf("\t%v\n", s)
+		}
+	}
+
+	if suggestionsMsg != "" {
+		requireMsg = fmt.Sprintf("%s\n%s", requireMsg, suggestionsMsg)
+		return fmt.Errorf(requireMsg, typedName, cmd.CommandPath())
+	}
+
+	return cmd.Help()
+}
+
+// suggestsByPrefixOrLd suggests a command by levenshtein distance or by prefix.
+// It returns an empty string if nothing was found
+func suggestsByPrefixOrLd(typedName, candidate string, minDistance int) string {
+	levenshteinDistance := ld(typedName, candidate, true)
+	suggestByLevenshtein := levenshteinDistance <= minDistance
+	suggestByPrefix := strings.HasPrefix(strings.ToLower(candidate), strings.ToLower(typedName))
+	if !suggestByLevenshtein && !suggestByPrefix {
+		return ""
+	}
+	return candidate
+}
+
+// ld compares two strings and returns the levenshtein distance between them.
+func ld(s, t string, ignoreCase bool) int {
```

Comment:
We may want to name the function a bit better, aka I don't mind "func levenshteinDistance" instead of `ld` (with makes me think of the `ld` command).

Reply:
actually in the cobra package they used the same name so that's why I kept it same :sweat_smile: